content/blog/angular-state-management-with-ngxs/index.md

```diff
@@ -0,0 +1,233 @@
+}
+```
+
+#### 4. Create your State:
+
+Here we have created the state of the todo, it contails the global state of the todo list.
+Create folder `state` and put the `todo-state.ts` file and add below code
+```typescript
+
+// File name todo-state.ts
+import { Injectable } from '@angular/core';
+import { State, NgxsOnInit, Action, StateContext } from '@ngxs/store';
+import { TodoActions } from './todo-actions';
+import { patch, updateItem } from '@ngxs/store/operators';
+
+export interface ITodo {
+    id: string;
+    title: string;
+    is_done: boolean;
+}
+export interface ITodoStateModel {
+    todoList: ITodo[];
+}
+@State<ITodoStateModel>({
+    name: 'todoList',
+    defaults: {
+        todoList: [],
+    },
+})
+@Injectable()
+export class TodoState implements NgxsOnInit {
+    ngxsOnInit(ctx) {
+        ctx.dispatch(new TodoActions.FetchAllTodos());
+    }
+    @Action(TodoActions.markDone)
+    markDone(
+      ctx: StateContext<ITodoStateModel>, 
+      { payload, is_done }: TodoActions.markDone
+    ) {
+        ctx.setState(
+            patch({
+                todoList: updateItem(
+                  (item: ITodo) => item.id === payload, 
+                  patch({ is_done: !is_done })
+                )
+            })
+        );
+    }
+
+    @Action(TodoActions.AddTodo)
+    add(
+        ctx: StateContext<ITodoStateModel>,
+        { payload }: TodoActions.AddTodo,
+    ) {
+        const state = ctx.getState();
+        ctx.setState({
+            ...state,
+            todoList: [
+                ...state.todoList,
+                {
+                    ...payload,
+                    id: Math.random().toString(36).substring(7),
+                    is_done: false
+                }
+            ],
+        }
+        );
+    }
+
+}
+```
+#### 5. Create your html view
+
+Here we have created a form which we use to create todo and listed all todo.
+put this code in your `app.component.html`
+```html
+<!-- File name  app.component.html -->
+<form [formGroup]="addForm" (ngSubmit)="onSubmit(addForm.value)">
+  <input type="text" formControlName="title" class="form-control todo-list-input" placeholder="What do you need to do today?">
+  <button class="add btn btn-primary font-weight-bold todo-list-add-btn">Add</button> 
+</form>
+<ul class="d-flex flex-column-reverse todo-list">
+    <li *ngFor="let todo of (todoList$ | async) ?. todoList" [ngClass]="{'completed': todo.is_done}">
+      <div class="form-check"> 
+        <label class="form-check-label"> 
+          <input (click)="markDone(todo.id, todo.is_done)" class="checkbox" type="checkbox" [checked]="todo.is_done">{{todo.title}} <i class="input-helper"></i></label> 
+      </div>
+    </li>
+  </ul>
+```
+Code code again code, now time to see the result so here is the UI you will get, to get the complete code please go [Github Repo](https://github.com/LoginRadius/engineering-blog-samples/tree/master/Angular/To-do-App-in-angular-with-NGXS)
```

Comment:
You can write something like `After lots of code, it's time to see the results. So here is the UI you will get. To get the complete code please go to [Github Repo](https://github.com/LoginRadius/engineering-blog-samples/tree/master/Angular/To-do-App-in-angular-with-NGXS)`